src/evaluators/ClassDeclaration.js

```diff
@@ -10,11 +10,272 @@
 /* @flow */
 
 import type { Realm } from "../realm.js";
-import type { LexicalEnvironment } from "../environment.js";
+import type { LexicalEnvironment, Reference } from "../environment.js";
 import type { Value } from "../values/index.js";
-import type { Reference } from "../environment.js";
-import type { BabelNodeClassDeclaration } from "babel-types";
+import { NullValue, EmptyValue, ObjectValue } from '../values/index.js';
+import type { BabelNodeClassDeclaration, BabelNodeExpression, BabelNodeClassMethod } from "babel-types";
+import parse from "../utils/parse.js";
+import {
+  HasOwnProperty,
+  SetFunctionName,
+  IsConstructor,
+  Get,
+  MakeConstructor,
+  CreateMethodProperty,
+  DefineMethod,
+  NewDeclarativeEnvironment,
+  MakeClassConstructor,
+  ObjectCreate,
+  ConstructorMethod,
+  GetValue,
+  IsStatic,
+  PropertyDefinitionEvaluation,
+  InitializeBoundName,
+  NonConstructorMethodDefinitions,
+} from "../methods/index.js";
+import { AbruptCompletion } from "../completions.js";
+import invariant from "../invariant.js";
 
+function EvaluateClassHeritage(realm: Realm, ClassHeritage: BabelNodeExpression, strictCode: boolean): ObjectValue | null {
+  let ref = realm.getRunningContext().lexicalEnvironment.evaluate(ClassHeritage, strictCode);
+  let val = GetValue(realm, ref);
+  if (!(val instanceof ObjectValue)) {
+    return null;
+  }
+  return val;
+}
+
+function ClassDefinitionEvaluation(realm: Realm, ast: BabelNodeClassDeclaration, className: string | void, strictCode: boolean, env: LexicalEnvironment) {
+  // 1. Let lex be the LexicalEnvironment of the running execution context.
+  let lex = env;
+
+  // 2. Let classScope be NewDeclarativeEnvironment(lex).
+  let classScope = NewDeclarativeEnvironment(realm, lex);
+
+  // 3. Let classScopeEnvRec be classScope’s EnvironmentRecord.
+  let classScopeEnvRec = classScope.environmentRecord;
+
+  // 4. If className is not undefined, then
+  if (className !== undefined) {
+    // a. Perform classScopeEnvRec.CreateImmutableBinding(className, true).
+    classScopeEnvRec.CreateImmutableBinding(className, true);
+  }
+
+  let protoParent;
+  let constructorParent;
+  // 5. If ClassHeritage opt is not present, then
+  if (ast.superClass === null) {
+    // a. Let protoParent be the intrinsic object %ObjectPrototype%.
+    protoParent = realm.intrinsics.ObjectPrototype;
+
+    // b. Let constructorParent be the intrinsic object %FunctionPrototype%.
+    constructorParent = realm.intrinsics.FunctionPrototype;
+  } else { // 6. Else
+    // a. Set the running execution context’s LexicalEnvironment to classScope.
+    realm.getRunningContext().lexicalEnvironment = classScope;
```

Comment:
As we propagate abrupt completions via exceptions, have a try-finally block around the code between setting and unsetting the .lexicalEnvironment.